tree/tree/src/TBranch.cxx

```diff
@@ -1128,7 +1128,7 @@ TBasket* TBranch::GetBasket(Int_t basketnumber)
    if (file == 0) {
       return 0;
    }
-   basket = GetFreshBasket();
+   basket = (fTree->GetMaxVirtualSize() < 0 || fTree->GetClusterPrefetch()) ? GetFreshCluster() : GetFreshBasket();
```

Comment:
Would it be more readable as
```
if (fTree->GetMaxVirtualSize() < 0 || fTree->GetClusterPrefetch())
   basket = GetFreshCluster();
else
   basket = GetFreshBasket();
```
?

Reply:
The description says:

> With MaxVirtualSize set to -1 **and** CacheDoClusterPrefetch true, there were 1.1 GB read in 90 read calls.

But the code says **or**.

Do we really need to test/involve GetMaxVirtualSize directly?

Reply:
Hi Philippe,
Cluster pre-fetch and retention are correlated.

Pre-fetching is controlled via fTree->GetClusterPrefetch and if it is on, second part of the if on line 1284 causes additional baskets of the cluster to be read via GetBasket(), but to make GetBasket not to re-use the first basket of the cluster, line 1131 has to call GetFreshCluster(), so we check for GetClusterPrefetch here as well.

Cluster retention is controlled by GetMaxVirtualSize and must be tested on line 1131 to call GetFreshCluster(), (otherwise GetFreshBasket() would re-use to be retained baskets), but we don't really need it on line 1284 when pre-fetching baskets. That means, if GetMaxVirtualSize is set to a negative value (ie, retaining clusters), only a single individual basket is read additional baskets of the same cluster are read on demand and data is retained on cluster boundary.

I am changing line 1284 and the test accordingly. David's measurements showed that for ATLAS best effects were seen when prefetching clusters and retaining one of them.

I will change line 1131 as you suggested to be more readable.

Thanks, Peter

Reply:
I see, I missed.

> By setting the MaxVirtualSize of the tree to a negative value, entire clusters will be loaded into memory and previous clusters will be retained - the absolute value of MaxVirtualSize indicates how many additional clusters will be kept in memory.

If this is the case, then why do we need GetClusterPrefetch? Wouldn't the negative value be enough of a trigger.

Although, this is still changing the semantic of MaxVirtualSize.  What I had in mind was:
* GetClusterPrefetch: control whether the per-cluster pre-load is done
* MaxVirtualSize, unchanged, control (indirectly) the number of cluster kept in memory by restricting that the sum of the size of the basket to be under that memory budget (and so to keep 2 cluster in memory you would need to set it to "size of cluster * compression_factor * 2)

Cheers,
Philippe.

Reply:
Sorry, that comment is obsolete and was our initial implementation, were both pre-fetching and retaining was controlled by negative MaxVirtualSize.
Now we were going to use negative MaxVirtualSize similar to negative TTreeCache size as control for cluster retaining:
* We agree on bullet 1: GetClusterPrefetch and that should be ok.
* Using unchanged MaxVirtualSize would mean we have to change the existing behavior. Currently once the limit is reached, TTree::DropBaskets() removes all Baskets from memory, rather than trimming off the oldest ones. Even for a large value (>size of cluster * compression_factor * 2) this would mean that when the third cluster starts reading, the first _and second_ cluster are deleted.
Rather than changing the existing implementation for positive, we preferred a negative value for MaxVirtualSize to trim off older cluster.
Does that make sense? Peter

Reply:
> Rather than changing ... we preferred ...

Yes, it is now clear.  I still need to weight the pros and cons of both choice ...